### PR TITLE
Add support for Uri to JsonSerializer

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/corefx/issues/38536.
Partially addresses https://github.com/dotnet/corefx/issues/38568.

This doesn't utilize the reader and writer because (Try)GetUri methods are not implemented yet: https://github.com/dotnet/corefx/issues/38647.
There are to-dos to replace serializer-level pre-processing once they are.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                value = default;
+                return false;
+            }
+
+            // TODO: use reader.GetUri() when https://github.com/dotnet/corefx/issues/38647 is implemented.
+            string uriString = reader.GetString();
+
+            try
+            {
+                value = new Uri(uriString);
```
Comment (watfordgnf) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Should we support relative URI's?
> This constructor assumes that the string parameter references an absolute URI and is equivalent to calling the Uri constructor with UriKind set to Absolute. If the string parameter passed to the constructor is a relative URI, this constructor will throw a UriFormatException.

As it stands it can write relative URI's but not read them.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class JsonValueConverterUri : JsonValueConverter<Uri>
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Can you hold off from committing this until https://github.com/dotnet/corefx/pull/38578 is in? There will be some minor refactoring. Thanks

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: This would be something the user would need to know statically \ up-front, correct? i.e. I don't think we want to try to dynamically figure out if a Uri is relative or not.

Assuming we don't add built-in support now, a custom converter could easily be authored to address relative URI's.

Built-in support would either create a second converter and make it public ("RelativeUriConverter") or make this public and add a bool property along with a custom attribute that can be applied to a property. In either case, this woud require an API review which is not feasible any longer for 3.0 due to schedule.

Comment (watfordgnf) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Would we not just use (`Ury.TryCreate(uriString, UriKind.RelativeOrAbsolute, out Uri tmp)`)[https://docs.microsoft.com/en-us/dotnet/api/system.uri.trycreate?view=netframework-4.8#System_Uri_TryCreate_System_String_System_UriKind_System_Uri__] ?

Comment (martincostello) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: I was about to suggest exactly the same thing, then it would support either.

Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Yes support both if it's that easy \ solution well-known.

I suppose, however, for some validation scenarios there are cases where you would want either a relative or absolute Uri and fail if not. That could be done through a custom converter.

Comment (martincostello) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: I think if someone did that it would be better to do that way, or custom code on the deserialized object, rather than throwing during the parsing which would be unrecoverable and hard to parse for validation reasons by application code.

Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Thanks, I utilized the `Uri.TryCreate` method as suggested.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class JsonConverterUri : JsonConverter<Uri>
+    {
+        public override Uri Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.String)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Technically checking token type like this is no longer required. The call to `GetString` below will cause the reader to throw JsonReaderException, which will be caught and re-thrown as JsonException.

File: src/System.Text.Json/tests/Serialization/TestClasses.SimpleTestClassWithObjectArrays.cs [hunk 4]
```diff
@@ -24,6 +24,7 @@ public class SimpleTestClassWithObjectArrays : ITestClass
         public object[] MySingle { get; set; }
         public object[] MyDouble { get; set; }
         public object[] MyDateTime { get; set; }
+        public object[] MyUri { get; set; }
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/TestClasses.SimpleTestClassWithObjectArrays.cs]: Please add some negative tests like `Parse<Uri>(@"""{}""")` and to make sure JsonException is being thrown from the converter.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+            writer.WriteStringValue(uriString);
+        }
+
+        public override void Write(Utf8JsonWriter writer, Uri value, JsonEncodedText propertyName, JsonSerializerOptions options)
+        {
+            // TODO: remove preprocessing when https://github.com/dotnet/corefx/issues/38647 is implemented.
+            string uriString = value.OriginalString;
+
+            if (uriString == string.Empty)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Why this special case? Can we add a test?

File: src/System.Text.Json/src/System.Text.Json.csproj [hunk 6]
```diff
@@ -79,6 +79,7 @@
     <Compile Include="System\Text\Json\Serialization\Converters\JsonValueConverterUInt16.cs" />
     <Compile Include="System\Text\Json\Serialization\Converters\JsonValueConverterUInt32.cs" />
     <Compile Include="System\Text\Json\Serialization\Converters\JsonValueConverterUInt64.cs" />
+    <Compile Include="System\Text\Json\Serialization\Converters\JsonValueConverterUri.cs" />
```
Comment (steveharter) [inline src/System.Text.Json/src/System.Text.Json.csproj]: FYI we eventually need to rename all of the files from "JsonValueConverter*" to just "JsonConverter*" to match file:classname. I added that task to [issue ](https://github.com/dotnet/corefx/issues/38408).

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs (hunk 5 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: Upon review, there's no need. Removed this check.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs [hunk 7]
```diff
@@ -27,27 +22,13 @@ public override Uri Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSeri
...
             // TODO: remove preprocessing when https://github.com/dotnet/corefx/issues/38647 is implemented.
-            string uriString = value.OriginalString;
-
-            if (uriString == string.Empty)
-            {
-                uriString = value.ToString();
-            }
-
-            writer.WriteStringValue(uriString);
+            writer.WriteStringValue(value.OriginalString);
```
Comment (watfordgnf) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: This will prevent relative URIs: ["This instance represents a relative URI, and this property is valid only for absolute URIs."](https://docs.microsoft.com/en-us/dotnet/api/system.uri.originalstring?view=netframework-4.8)

(I didn't realize how complicated the URI object was w.r.t. absolute vs relative URI's until just now.)

Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterUri.cs]: `OriginalString` seems to work for both relative and absolute -

```
Uri.TryCreate("https://domain/path", UriKind.Absolute, out Uri uri1);
Console.WriteLine(uri1.OriginalString); // https://domain/path
Console.WriteLine(uri1.ToString()); // https://domain/path
		
Uri.TryCreate("~/path", UriKind.Relative, out Uri uri2);
Console.WriteLine(uri2.OriginalString); // ~/path
Console.WriteLine(uri2.ToString()); // ~/path
```

Per documentation - `ToString()` seems the safest 
(no exceptions) for now till this support is properly implemented in the reader/writer - https://github.com/dotnet/corefx/issues/38647.

**Comment by layomia** [comment]
Failing tests are unrelated - System.Net.Http.Functional.Tests

File: src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs [hunk 8]
```diff
@@ -41,13 +41,6 @@ public static void ThrowJsonException_DeserializeUnableToConvertValue(Type prope
             throw new JsonException(message, path, null, null, innerException);
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        public static JsonException GetJsonException_DeserializeUnableToConvertValue(Type propertyType, string path)
-        {
-            string message = SR.Format(SR.DeserializeUnableToConvertValue, propertyType) + $" Path: {path}.";
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: I wonder why these exceptions/messages are no longer relevant?

File: src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs [hunk 9]
```diff
@@ -145,11 +126,6 @@ public static void ThrowJsonException_DeserializeDataRemaining(long length, long
             throw new JsonException(SR.Format(SR.DeserializeDataRemaining, length, bytesRemaining), path: null, lineNumber: null, bytePositionInLine: null);
         }
 
-        public static void ThrowJsonException_DeserializeDuplicateKey(string key, in Utf8JsonReader reader, string path)
-        {
-            ThrowJsonException(SR.Format(SR.DeserializeDuplicateKey, key), in reader, path);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: Does anyone still use `SR.DeserializeDuplicateKey`?
